File: document/documentation/reference/operation/focal/convolve/example-1/convolve.cpp
#include "lue/framework/api/cxx.hpp"
#include "lue/document.hpp"
#include <hpx/hpx_main.hpp>


class Example: public lue::document::Example
{

    public:

        Example(int const argc, char const* const* argv):

            lue::document::Example{
                [&]() -> cxxopts::Options
                {
                    cxxopts::Options options{std::filesystem::path{argv[0]}.filename()};

                    options.positional_help("<array> <kernel> <result>");
                    options.add_options()
                        // clang-format off
                        ("argument_array", "array to read", cxxopts::value<std::string>())
                        ("argument_kernel", "kernel to read", cxxopts::value<std::string>())
                        ("result_array", "array to write", cxxopts::value<std::string>())
                        // clang-format on
                        ;
                    options.parse_positional({"argument_array", "argument_kernel", "result_array"});

                    return options;
                }(),
                argc,
                argv}

        {
        }

    private:

        auto run_implementation() -> int override
        {
            if (!argument_parsed("argument_array") || !argument_parsed("argument_kernel") ||
                !argument_parsed("result_array"))
            {
                throw std::runtime_error("Wrong usage, pass '--help' for more information");
            }

            auto const argument_array_pathname = argument<std::string>("argument_array");
            auto const argument_kernel_pathname = argument<std::string>("argument_kernel");
            auto const result_array_pathname = argument<std::string>("result_array");

            using namespace lue::api;

            // [example
            Field const array = from_gdal(argument_array_pathname);
            Kernel const kernel = lue::document::read_kernel(argument_kernel_pathname);
            Field const result = convolve(array, kernel);

            to_gdal(result, result_array_pathname);
            // example]

            return EXIT_SUCCESS;
        }
};


auto main(int argc, char** argv) -> int
{
    return Example{argc, argv}.run();
}

File: document/documentation/reference/operation/focal/convolve/example-1/convolve.py
import sys
from pathlib import Path

import docopt

import lue.document as ld
import lue.framework as lfr
from lue import __version__ as version


@lfr.runtime_scope
def main() -> None:
    command = Path(sys.argv[0]).name
    usage = f"""\
Usage:
    {command} <array> <kernel> <result>

Options:
    -h --help   Show this screen and exit
    --version   Show version and exit
    <array>     Array to read
    <kernel>    Kernel to read
    <result>    Array to write
"""
    arguments = docopt.docopt(usage, sys.argv[1:], version=version)
    argument_array_path = Path(arguments["<array>"])
    argument_kernel_path = Path(arguments["<kernel>"])
    result_array_path = Path(arguments["<result>"])

    array = ld.read_array(argument_array_path)
    kernel = ld.read_kernel(argument_kernel_path)

    # [example
    result = lfr.convolve(array, kernel)
    # example]

    lfr.to_gdal(result, str(result_array_path))


if __name__ == "__main__":
    sys.exit(main())
